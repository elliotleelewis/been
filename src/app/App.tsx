import { type FC } from 'react';

import { CountriesProvider } from './contexts/countries-context';
import { Header } from './header/header';
import { Map } from './map/map';
import { Menu } from './menu/menu';

export const App: FC = () => {
	return (
		<CountriesProvider>
			<div className="flex h-full flex-col-reverse sm:flex-row">
				<div className="flex basis-1/3 flex-col overflow-auto dark:bg-zinc-900 dark:text-white">
					<Menu header={<Header show="tablet" />} />
				</div>
				<div className="basis-2/3">
					<Map header={<Header />} />
				</div>
			</div>
		</CountriesProvider>
	);
};
