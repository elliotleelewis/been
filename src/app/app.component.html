<mgl-map
	[attributionControl]="false"
	[minZoom]="minZoom"
	[style]="'mapbox://styles/mapbox/streets-v9'"
	[zoom]="[minZoom]"
></mgl-map>
