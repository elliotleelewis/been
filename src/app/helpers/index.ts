export * from './coordinates';
export * from './regionalizer';
