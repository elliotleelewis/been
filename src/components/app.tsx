import classNames from 'classnames';
import { useSetAtom } from 'jotai';
import { memo, useCallback, useEffect, useState } from 'react';
import type { FC } from 'react';
import { rawCountriesAtom } from '../state/atoms.ts';
import { Globe } from './globe';
import { Menu } from './menu';

export const App: FC = memo(() => {
	const setRawCountries = useSetAtom(rawCountriesAtom);
	const [loading, setLoading] = useState(true);
	const [error, setError] = useState<Error | null>();

	const [menuFullscreen, setMenuFullscreen] = useState(false);

	useEffect(() => {
		import('../data/countries')
			.then(({ countries }) => {
				const countryMap = Object.fromEntries(
					countries.map((c) => [c.iso3166, c]),
				);
				setRawCountries(countryMap);
				setLoading(false);
			})
			.catch((error) => {
				setError(error);
			});
	}, [setRawCountries]);

	const reload = useCallback(() => {
		window.location.reload();
	}, []);

	const toggleMenuFullscreen = useCallback(() => {
		setMenuFullscreen((val) => !val);
	}, []);

	return (
		<div className="grid size-full grid-rows-[auto,1fr,auto] md:grid-cols-3 md:grid-rows-[auto,1fr] dark:bg-zinc-900 dark:text-white">
			<div className="flex items-center justify-center bg-primary p-3 text-white md:col-span-1 dark:bg-zinc-950 dark:text-primary">
				<h1 className="select-none font-bold text-xl tracking-wide">
					been
				</h1>
			</div>
			<div
				className={classNames(
					'z-10 order-3 flex flex-col overflow-auto md:order-2 md:col-span-1 md:row-start-2 dark:bg-zinc-900 dark:text-white',
					menuFullscreen && '-mt-[60vh] md:mt-0',
				)}
			>
				{error ? (
					<div className="flex size-full flex-col items-center justify-center gap-2 px-2 text-center text-lg">
						<span>
							Oops! Something went wrong whilst loading the list
							of countries.
						</span>
						<button
							type="button"
							className="items-center justify-center rounded-md border border-primary px-6 py-2 text-primary transition focus:ring-2 focus:ring-primary/50 active:ring-primary"
							onClick={reload}
						>
							Try again
						</button>
					</div>
				) : (
					<Menu
						loading={loading}
						fullscreen={menuFullscreen}
						toggleFullscreen={toggleMenuFullscreen}
					/>
				)}
			</div>
			<div className="order-2 min-h-[60vh] md:order-3 md:col-span-2 md:row-span-2">
				<Globe />
			</div>
		</div>
	);
});
App.displayName = 'App';
