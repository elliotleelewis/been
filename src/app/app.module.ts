import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { BrowserModule } from '@angular/platform-browser';
import { MAPBOX_API_KEY, NgxMapboxGLModule } from 'ngx-mapbox-gl';

import { ENVIRONMENT } from '../environments/environment';

import { AppComponent } from './app.component';
import { HeaderComponent } from './header/header.component';
import { MapComponent } from './map/map.component';
import { MenuComponent } from './menu/menu.component';

@NgModule({
	declarations: [AppComponent, HeaderComponent, MapComponent, MenuComponent],
	imports: [
		BrowserModule,
		FormsModule,
		ReactiveFormsModule,
		NgxMapboxGLModule,
	],
	providers: [
		{
			provide: MAPBOX_API_KEY,
			useValue: ENVIRONMENT.apiKeyMapbox,
		},
	],
	bootstrap: [AppComponent],
})
export class AppModule {}
