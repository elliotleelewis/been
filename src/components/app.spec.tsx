import { render } from '@testing-library/react';
import { beforeAll, describe, expect, test, vi } from 'vitest';

import { App } from './app';

describe('App', () => {
	beforeAll(() => {
		vi.spyOn(window, 'matchMedia').mockImplementation(
			() =>
				({
					matches: false,
					addEventListener: vi.fn(),
					removeEventListener: vi.fn(),
				}) satisfies Partial<MediaQueryList> as unknown as MediaQueryList,
		);
	});

	test('it should render', () => {
		const result = render(<App data={[]} />);

		expect(result).toMatchSnapshot();
	});
});
