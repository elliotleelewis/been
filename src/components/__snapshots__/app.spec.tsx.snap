// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > should render 1`] = `
<DocumentFragment>
  <div
    class="grid size-full grid-rows-[auto,1fr,auto] md:grid-cols-3 md:grid-rows-[auto,1fr] dark:bg-zinc-900 dark:text-white"
  >
    <div
      class="flex items-center justify-center bg-primary p-3 text-white md:col-span-1 dark:bg-zinc-950 dark:text-primary"
    >
      <h1
        class="select-none font-bold text-xl tracking-wide"
      >
        been
      </h1>
    </div>
    <div
      class="z-10 order-3 flex flex-col overflow-auto md:order-2 md:col-span-1 md:row-start-2 dark:bg-zinc-900 dark:text-white"
    >
      <div
        class="flex border-zinc-200 border-t-2 sm:border-t-0 dark:border-zinc-700"
      >
        <form
          class="w-full"
        >
          <label
            class="sr-only"
            for="«r0»"
          >
            Search
          </label>
          <input
            class="w-full border-none bg-white px-4 py-2 text-zinc-900 focus:ring-2 focus:ring-primary/50 focus:ring-inset disabled:hover:cursor-not-allowed dark:bg-zinc-900 dark:text-white"
            disabled=""
            id="«r0»"
            placeholder="Search..."
            type="text"
            value=""
          />
        </form>
        <button
          class="visible p-2 md:hidden"
          type="button"
        >
          <svg
            class="size-4"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title>
              Expand
            </title>
            <path
              d="M3.75 3.75v4.5m0-4.5h4.5m-4.5 0L9 9M3.75 20.25v-4.5m0 4.5h4.5m-4.5 0L9 15M20.25 3.75h-4.5m4.5 0v4.5m0-4.5L15 9m5.25 11.25h-4.5m4.5 0v-4.5m0 4.5L15 15"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </button>
      </div>
      <div
        class="flex size-full items-center justify-center"
      >
        <svg
          class="size-5 animate-spin text-neutral-800 dark:text-white"
          fill="none"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title>
            Loading!
          </title>
          <circle
            class="opacity-25"
            cx="12"
            cy="12"
            r="10"
            stroke="currentColor"
            stroke-width="4"
          />
          <path
            class="opacity-75"
            d="M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
    <div
      class="order-2 min-h-[60vh] md:order-3 md:col-span-2 md:row-span-2"
    >
      <div
        style="position: relative; width: 100%; height: 100%;"
      />
    </div>
  </div>
</DocumentFragment>
`;
